pkg/providers/aws/provider_postgressnapshot.go

```diff
@@ -0,0 +1,244 @@
+
+	if err != nil {
+		errMsg := "failed to create AWS session"
+		return croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	rdsSvc := rds.New(session)
+
+	return p.deletePostgresSnapshot(ctx, snapshot, postgres, rdsSvc)
+}
+
+func (p *PostgresSnapshotProvider) createPostgresSnapshot(ctx context.Context, snapshot *v1alpha1.PostgresSnapshot, postgres *v1alpha1.Postgres, rdsSvc rdsiface.RDSAPI) (*providers.PostgresSnapshotInstance, croType.StatusMessage, error) {
+	// generate snapshot name
+	snapshotName, err := BuildTimestampedInfraNameFromObjectCreation(ctx, p.client, snapshot.ObjectMeta, DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to generate snapshot name"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// update cr with snapshot name
+	snapshot.Status.SnapshotID = snapshotName
+
+	if err = p.client.Status().Update(ctx, snapshot); err != nil {
+		errMsg := fmt.Sprintf("failed to update instance %s in namespace %s", snapshot.Name, snapshot.Namespace)
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// get instance name
+	instanceName, err := BuildInfraNameFromObject(ctx, p.client, postgres.ObjectMeta, DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to get cluster name"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	foundSnapshot, err := p.findSnapshotInstance(rdsSvc, snapshotName)
+
+	if err != nil {
+		errMsg := "failed to describe snaphots in AWS"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// create snapshot of the rds instance
+	if foundSnapshot == nil {
+		// postgres instance has either just been created
+		// or is already backing up
+		if postgres.Status.Phase == croType.PhaseInProgress {
+			return nil, croType.StatusMessage("waiting for postgres instance to be available"), nil
+		}
+		// postgres instance is being deleted
+		// impossible to create a snapshot
+		if postgres.Status.Phase == croType.PhaseDeleteInProgress {
+			errMsg := "cannot create snapshot when instance deletion is in progress"
+			return nil, croType.StatusMessage(errMsg), errorUtil.New(errMsg)
+		}
+		p.logger.Info("creating rds snapshot")
+		_, err = rdsSvc.CreateDBSnapshot(&rds.CreateDBSnapshotInput{
+			DBInstanceIdentifier: aws.String(instanceName),
+			DBSnapshotIdentifier: aws.String(snapshotName),
+		})
+		if err != nil {
+			errMsg := "error creating rds snapshot"
+			return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+		}
+		return nil, "snapshot started", nil
+	}
+
+	// if snapshot status complete update status
+	if *foundSnapshot.Status == "available" {
+		return &providers.PostgresSnapshotInstance{
+			Name: *foundSnapshot.DBSnapshotIdentifier,
+		}, "snapshot created", nil
+	}
+
+	// creation in progress
+	msg := fmt.Sprintf("current snapshot status : %s", *foundSnapshot.Status)
+	p.logger.Info(msg)
```

Comment:
Just a small one, we are in the process of updating the logging format in CRO - https://github.com/integr8ly/cloud-resource-operator/blob/master/pkg/providers/aws/provider_postgres.go#L105-L106

Using the `withField` and `action` we can pin point the function that called the log. Like adding the developer docs we are trying to make a concious effort to update these as we go, to avoid large refactors/tech debt 

There is even a utility function for it I just noticed :) https://github.com/integr8ly/cloud-resource-operator/blob/master/pkg/resources/logger.go#L9